DOC: #22896, Fixed docstring of to_stata in pandas/core/frame.py

ryankarlos opened the pull request:
- [x] closes #22896
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] passes `./scripts/validate_docstrings.py pandas.DataFrame.to_stata`

pep8speaks commented:
Hello @ryankarlos! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/frame.py`](https://github.com/ryankarlos/pandas/blob/201605170f429d10c2f7c35ec587a5576c670f4d/pandas/core/frame.py) !

 - There are no PEP8 issues in the file [`pandas/core/strings.py`](https://github.com/ryankarlos/pandas/blob/201605170f429d10c2f7c35ec587a5576c670f4d/pandas/core/strings.py) !

 - Complete extra results for this file :

> file_to_check.py:939:-759: W605 invalid escape sequence '\d'
file_to_check.py:939:-612: W605 invalid escape sequence '\d'
file_to_check.py:939:-441: W605 invalid escape sequence '\d'
file_to_check.py:939:-239: W605 invalid escape sequence '\d'
file_to_check.py:939:-78: W605 invalid escape sequence '\d'
file_to_check.py:1022:-782: W605 invalid escape sequence '\d'
file_to_check.py:1022:-596: W605 invalid escape sequence '\d'
file_to_check.py:1022:-379: W605 invalid escape sequence '\d'
file_to_check.py:1022:-144: W605 invalid escape sequence '\d'
---

ryankarlos commented:
Ive followed the docstring format here https://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html 

Ive added in a returns section, corrected linting issues and modified the examples as some of the variables (mainly for the date example) were not defined and hence failing some of the tests. 

For some reason, the tests were throwing up linting errors for an unrelated pull request - but Ive corrected these as well.

ryankarlos commented:
This is actually my fault - i had not cleared my previous git history --- if you want me to resubmit then let me know

datapythonista commented:
Thanks for the work on this @ryankarlos 

Can you leave in this PR just the changes in the `to_stata` docstring please? Things become more complex if we mix things, and I see some blank lines added in unrelated code, and changes to the `split` function (which are welcome, but in a separate PR).

I also see that there are PEP-8 issues in the examples, and may be they can be simplified a bit.

datapythonista commented on pandas/core/frame.py:
```diff
@@ -1911,26 +1915,32 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
+        >>> dates_1 = pd.to_datetime(["2016-05-06","2017-05-07", "2018-05-10"])
+        >>> dates_2 = pd.to_datetime(["2016-11-06","2017-11-07", "2018-11-10"])
+        >>> df_dates = pd.DataFrame({'date1': dates_1, 'date2': dates_2})
+        >>> df_dates.to_stata('./date_data_file.dta', {2 : 'tw'})
```
Can you explain what the `{2: 'tw'}` means? I know this is part of the original docstring, but I think we should make the examples in a way that they explain what is going on. I don't know much about stata or this function, and this without explanation is not useful to me.

Also, would be great to simplify this. May be just a date column if that makes sense, without creating variables first, and may be using `date_range` to avoid so much information that does not add value. And if we can create a single DataFrame at the beginning that is useful for all functions, that would be much better.

Finally, if we can use something looking more like a real example. I think it gives a better impression and helps understand than a [1, 2, 3]. We've got some examples with animals in `pandas/core/generic.py`, I'd use one of them.

ryankarlos replied:
The `tw` is a Stata specific option for displaying time in weeks i believe - i have updated it with  a single data frame using the animals examples from `pandas/core/generic.py` and excluded the separate example specifically for dates  (i agree it does not really add any value).

datapythonista commented on pandas/core/frame.py:
```diff
@@ -1911,26 +1915,32 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
 
-        >>> writer = StataWriter('./data_file.dta', data)
+        >>> StataWriter = pd.io.stata.StataWriter
+        >>> writer = StataWriter('./data_file.dta', df)
```
I'd use instead (assuming you're using the animals data):
`>>> writer = pd.io.stata.StataWriter('animals.dta', df)`

ryankarlos commented:
just modified the changes

ryankarlos committed:
Updated examples

ryankarlos commented:
> Thanks for the work on this @ryankarlos
> 
> Can you leave in this PR just the changes in the `to_stata` docstring please? Things become more complex if we mix things, and I see some blank lines added in unrelated code, and changes to the `split` function (which are welcome, but in a separate PR).
> 
> I also see that there are PEP-8 issues in the examples, and may be they can be simplified a bit.

Yes sorry, I have undone the changes to the other docstring and only left the to_stata one. I wasn't sure what those PEP-8 issues regarding the escape sequence were referring to - I have modified and simplified the dates in the example to include `-` instead of `\`  .... not sure if that was the problem.

ryankarlos committed:
Add returns section

ryankarlos committed:
Fix whitespace on blank line 1918:1: W293

ryankarlos commented:
@datapythonista let me know if there are any updates required

datapythonista commented:
Looks good, added some comments.

ryankarlos commented:
@datapythonista Did you add comments to the latest commit - as I cannot find anything ?

datapythonista reviewed (commented):
Sorry, I don't have a computer right now, and it's tricky to do the reviews from the phone. I think the comments should be added now.

ryankarlos committed:
Incorporated comments

ryankarlos commented:
Thanks for the comments - incorporated the changes.

datapythonista reviewed (commented):
Sorry @ryankarlos, I've just seen that I didn't submit the comments of the last review (I did it from the phone, and it's tricky).

Besides those, can you edit `ci/code_checks.py` and in the part of the doctests, stop excluding this docstring, as they should now pass.

Thanks!

ryankarlos committed:
Minor changes

ryankarlos commented:
@datapythonista  added requested changes

datapythonista reviewed (approved):
Looks good. Can you add `optional` to the `data_label` parameter, and the default to `version : {114, 117}, default 114`.

I think with that is ready to be merged (if you can also take a look at the CI and make sure it's green, so can merge).

ryankarlos committed:
Final

ryankarlos commented:
@datapythonista Thanks, changes added -  all green now

datapythonista commented:
lgtm, just changed a word. I'll wait for for someone else to merge this, so we have an extra pair of eyes on it.

Thanks @ryankarlos

jreback reviewed (changes requested):
tiny comments. @bashtage if you can have a look.

ryankarlos committed:
Final fixes

ryankarlos commented on pandas/core/frame.py:
```diff
@@ -1921,28 +1924,17 @@ def to_stata(self, fname, convert_dates=None, write_index=True,
             files.
 
-        Examples
+        Exampless
```
@jreback done. Also fixed some merge conflicts after rebasing with master.

ryankarlos commented:
@jreback  All green now

jreback commented:
thanks @ryankarlos
